### PR TITLE
IDE-674 ESDL files with *.ECM extension should be treated as ESDL code

### PR DESCRIPTION
**Pull request by dehilsterlexis**
Fixed IDE-674

Signed-off-by: dehilsterlexis <david.dehilster@lexisnexis.com>

### CODE REVIEW

**Comment by dehilsterlexis** [comment]
@JamesDeFabia @rpastrana please test ESDL and ECM files with the zipped debug version of the IDE in this comment.
[ECLIDE.zip](https://github.com/hpcc-systems/eclide/files/1431357/ECLIDE.zip)

**Comment by dehilsterlexis** [comment]
@JamesDeFabia @rpastrana ping

**Comment by dehilsterlexis** [comment]
@GordonSmith ready for code review

**Comment by GordonSmith** [comment]
Can you re-target this for candidate-6.4.6 (via the edit button at the top right of this page).

**Comment by richardkchapman** [comment]
Also please fix the typo in the commit message title

**Comment by dehilsterlexis** [comment]
@richardkchapman ready

**Comment by richardkchapman** [comment]
@dehilsterlexis  You edited the PR title in github but the commit is still wrong. You need to do a git commit -s --amend, then a git push --force.

**Comment by dehilsterlexis** [comment]
@richardkchapman sorry about that. Done!

**Comment by richardkchapman** [comment]
@GordonSmith have you reviewed this?

File: comms/AttributeImpl.h [hunk 1]
```diff
@@ -191,7 +196,7 @@ class CAttributeBase : public clib::CLockableUnknown
 
             CComPtr<IRepository> rep = AttachModFileRepository(outputFile.TempFileName(), false);
             outputFile.HandsOn();
-            if (action == PREPROCESS_SUBMIT && metaInfo[MetaInfoItemRepositoryCode] == ATTRIBUTE_TYPE_ESDL)
+            if (action == PREPROCESS_SUBMIT && (metaInfo[MetaInfoItemRepositoryCode] == ATTRIBUTE_TYPE_ESDL || metaInfo[MetaInfoItemRepositoryCode] == ATTRIBUTE_TYPE_ECM))
```
Comment (GordonSmith) [inline comms/AttributeImpl.h]: Is the second half of the OR actually needed, given your code change above?

**Comment by rpastrana** [comment]
Tried to test, but encountered a crash. will re-test when @dehilsterlexis send updated private build.

**Comment by dehilsterlexis** [comment]
@rpastrana I will email it to you.

**Comment by dehilsterlexis** [comment]
@GordonSmith you were correct. The checking for ECM type in comms/AttributeImpl.h at that point was unnecessarily. Back to you.

File: comms/AttributeType.cpp [hunk 2]
```diff
@@ -169,8 +169,11 @@ class CAttributeType : public IAttributeType, public CUnknown
     {
         if (m_fileExtension.empty())
         {
-            m_fileExtension = _T(".");
-            m_fileExtension += GetRepositoryCode();
+            m_fileExtension = GetRepositoryCode();
+        }
+        if (!boost::istarts_with(m_fileExtension, _T(".")))
```
Comment (GordonSmith) [inline comms/AttributeType.cpp]: Are there "repository codes" which contain a "."?

File: comms/AttributeType.cpp [hunk 3]
```diff
@@ -250,8 +253,10 @@ static CacheT<std::_tstring, CAttributeType> AttributeTypeCache;
 
 IAttributeType *AttributeTypeFromExtension(const std::_tstring & extension) {
     IAttributeType *attrType = NULL;
-    if (boost::algorithm::iequals(extension, ATTRIBUTE_TYPE_ESDL) || boost::algorithm::iequals(extension, ATTRIBUTE_TYPE_ECM))
+    if (boost::algorithm::iequals(extension, ATTRIBUTE_TYPE_ESDL))
         attrType = CreateIAttributeESDLType();
+    else if (boost::algorithm::iequals(extension, ATTRIBUTE_TYPE_ECM))
+        attrType = CreateIAttributeECMType();
```
Comment (GordonSmith) [inline comms/AttributeType.cpp]: I thought we were trying to not use the ECM type now?

File: comms/ModFileRepository.cpp [hunk 4]
```diff
@@ -419,7 +419,7 @@ class CModFileRepository : public IRepository, public clib::CLockableUnknown
     IAttribute * InsertAttribute(const TCHAR* module, const TCHAR* attribute, IAttributeType * type) const
     {
         clib::recursive_mutex::scoped_lock proc(m_mutex);
-        return CreateModFileAttribute(this, module, attribute, type->GetFileExtension(), _T(""));
+        return CreateModFileAttribute(this, module, attribute, type->GetRepositoryCode(), _T(""));
```
Comment (GordonSmith) [inline comms/ModFileRepository.cpp]: I think this one is meant to be the extension?

**Comment by richardkchapman** [comment]
@dehilsterlexis

File: comms/AttributeType.cpp (hunk 3 above)
Comment (dehilsterlexis) [reply comms/AttributeType.cpp]: @GordonSmith I did think a lot about this and in my changes to treat them the same, I thought it cleaner to keep them as separate types but in the end, treat them the same. The problem is having an extension and a type that doesn't match. I think this cleaner.

**Comment by dehilsterlexis** [comment]
@GordonSmith I changed the function call back to Extension as you pointed out in InsertAttribute in ModFilefRepository line 422. Back to you.

**Comment by dehilsterlexis** [comment]
@GordonSmith

**Comment by dehilsterlexis** [comment]
@richardkchapman ready to merge

**Pull request merged**